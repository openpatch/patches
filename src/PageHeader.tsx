import styled from "@emotion/styled";
import { ReactElement, ReactNode } from "react";
import { Box } from "./Box";
import { ButtonGroup } from "./ButtonGroup";
import { ButtonOutlineProps } from "./ButtonOutline";
import { ButtonPrimaryProps } from "./ButtonPrimary";
import { ButtonSecondaryProps } from "./ButtonSecondary";
import { Heading, HeadingProps } from "./Heading";
import { variant } from "./system";
import { SystemColor, SystemShadows, SystemSizes } from "./system/types";

type ActionProps =
  | ButtonPrimaryProps
  | ButtonSecondaryProps
  | ButtonOutlineProps;

export type PageHeaderProps = {
  children?: ReactNode | string;
  variant?: "stack" | "overlap";
  lineClamp?: HeadingProps["lineClamp"];
  actions?: ReactElement<ActionProps> | ReactElement<ActionProps>[];
  meta?: ReactNode;
};

const StyledHeader = styled(Box)(
  variant<
    {
      color?: SystemColor;
      backgroundColor?: SystemColor;
      boxShadow?: SystemShadows;
      marginTop?: SystemSizes | SystemSizes[];
      "& > div"?: object;
    },
    "stack" | "overlap",
    string
  >({
    scale: "pageHeader",
    prop: "variant",
    variants: {
      stack: {
        backgroundColor: "card",
        boxShadow: "standard",
      },
      overlap: {
        backgroundColor: "primary.900",
        color: "primary.50",
        "& > div": {
          borderTopColor: "primary.800",
          borderTopStyle: "solid",
          borderTopWidth: "standard",
          paddingTop: ["medium", "large"],
        },
      },
    },
  })
);

export const PageHeader = ({
  children,
  lineClamp = 2,
  variant = "stack",
  actions,
  meta,
}: PageHeaderProps) => {
  return (
    <StyledHeader as="header" variant={variant}>
      <Box
        maxWidth="large"
        margin="auto"
        display="flex"
        alignItems="center"
        justifyContent="space-between"
        paddingY={["xxsmall", "small"]}
        px={["small", "medium", "medium"]}
        flexWrap="wrap"
        gap="standard"
      >
        <Box display="flex" flexDirection="column">
          <Box mr={["small", "medium", "medium"]}>
            {typeof children === "string" ? (
              <Heading
                title={children}
                lineClamp={lineClamp}
                as="h1"
                fontSize="xlarge"
              >
                {children}
              </Heading>
            ) : (
              children
            )}
          </Box>
          {meta && <Box>{meta}</Box>}
        </Box>
        {actions && <ButtonGroup space="xsmall">{actions}</ButtonGroup>}
      </Box>
    </StyledHeader>
  );
};
