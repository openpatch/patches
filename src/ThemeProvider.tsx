import React, { ReactElement } from "react";
import { ThemeProvider as EThemeProvider } from "emotion-theming";
import { Theme, base } from "./themes";
import { GlobalCssTheme } from "./themes/utils";

interface ThemeProviderProps {
  theme?: Theme;
  children: ReactElement;
}

const ThemeProvider = ({ theme = base, children }: ThemeProviderProps) => {
  return (
    <EThemeProvider theme={theme}>
      <GlobalCssTheme theme={theme} />
      {children}
    </EThemeProvider>
  );
};

export default ThemeProvider;
