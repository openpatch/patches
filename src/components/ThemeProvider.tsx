import React, { ReactElement } from "react";
import {
  ThemeProvider as SCThemeProvider,
  DefaultTheme,
} from "styled-components";
import defaultTheme from "../shared/theme";

interface ThemeProviderProps {
  theme?: DefaultTheme;
  children: ReactElement;
}

const ThemeProvider = ({
  theme = defaultTheme,
  children,
}: ThemeProviderProps) => {
  return <SCThemeProvider theme={theme}>{children}</SCThemeProvider>;
};

export default ThemeProvider;
