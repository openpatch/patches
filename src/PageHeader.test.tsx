import React from "react";
import { render } from "@testing-library/react";
import { PageHeader, PageHeaderProps } from "./PageHeader";
import { ThemeProvider } from "./ThemeProvider";

const renderWithTheme = (props?: PageHeaderProps) =>
  render(
    <ThemeProvider>
      <PageHeader {...props} />
    </ThemeProvider>
  );

test("should render", () => {
  renderWithTheme();
});

test("should match snapshot", () => {
  const { container } = renderWithTheme();
  expect(container).toMatchSnapshot();
});
