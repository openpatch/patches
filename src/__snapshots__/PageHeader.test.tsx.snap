// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.emotion-0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  background-color: #fff;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1),0 1px 2px 0 rgba(0, 0, 0, 0.06);
}

.emotion-1 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  max-width: 1280px;
  margin: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 16px;
  padding-right: 16px;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  gap: 16px;
}

@media screen and (min-width: 768px) {
  .emotion-1 {
    padding-top: 16px;
    padding-bottom: 16px;
    padding-left: 24px;
    padding-right: 24px;
  }
}

@media screen and (min-width: 1024px) {
  .emotion-1 {
    padding-left: 24px;
    padding-right: 24px;
  }
}

.emotion-2 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-3 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  margin-right: 16px;
}

@media screen and (min-width: 768px) {
  .emotion-3 {
    margin-right: 24px;
  }
}

@media screen and (min-width: 1024px) {
  .emotion-3 {
    margin-right: 24px;
  }
}

<div>
  <div
    id="patches-root"
  >
    <header
      class="emotion-0"
    >
      <div
        class="emotion-1"
        display="flex"
      >
        <div
          class="emotion-2"
          display="flex"
        >
          <div
            class="emotion-3"
          />
        </div>
      </div>
    </header>
  </div>
</div>
`;
